src/runtime/disco/threaded_session.cc

```diff
@@ -0,0 +1,184 @@
+ public:
+  void Send(const TVMArgs& args) {
+    RPCReference::ReturnPackedSeq(args.values, args.type_codes, args.num_args, this);
+    NotifyEnqueue();
+  }
+
+  TVMArgs Recv() {
+    WaitDequeue();
+    uint64_t packet_nbytes = 0;
+    RPCCode code = RPCCode::kReturn;
+    this->Read(&packet_nbytes);
+    this->Read(&code);
+    TVMValue* values = nullptr;
+    int* type_codes = nullptr;
+    int num_args = 0;
+    RPCReference::RecvPackedSeq(&values, &type_codes, &num_args, this);
+    return TVMArgs(values, type_codes, num_args);
+  }
+
+ protected:
+  void NotifyEnqueue() {
+    {
+      std::lock_guard<std::mutex> lock{mutex_};
+      ++msg_cnt_;
+    }
+    condition_.notify_one();
+  }
+
+  void WaitDequeue() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    condition_.wait(lock, [this] { return msg_cnt_.load() > 0; });
+    --msg_cnt_;
+  }
+
+  void MessageStart(uint64_t packet_nbytes) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    size_t n = ring_buffer_.bytes_available();
+    n += packet_nbytes + sizeof(uint64_t);
+    this->ring_buffer_.Reserve(n);
+  }
+
+  void MessageDone() {}
+
+  void ThrowError(RPCServerStatus status) {
+    LOG(FATAL) << "InternalError: Unexpected error in RPC: " << RPCServerStatusToString(status);
+  }
+
+  template <typename T>
+  T* ArenaAlloc(int count) {
+    static_assert(std::is_pod<T>::value, "need to be trival");
+    return arena_.template allocate_<T>(count);
+  }
+
+  size_t Read(void* data, size_t size) final {
+    std::lock_guard<std::mutex> lock(mutex_);
+    ring_buffer_.Read(data, size);
+    return size;
+  }
+
+  void Write(const void* data, size_t size) final {
+    std::lock_guard<std::mutex> lock(mutex_);
+    ring_buffer_.Write(data, size);
+  }
+
+  using dmlc::Stream::Read;
+  using dmlc::Stream::ReadArray;
+  using dmlc::Stream::Write;
+  using dmlc::Stream::WriteArray;
+  friend struct RPCReference;
+
+  std::mutex mutex_;
+  std::atomic<int> msg_cnt_{0};
+  std::condition_variable condition_;
+
+  support::RingBuffer ring_buffer_;
+  support::Arena arena_;
+};
+
+class DiscoThreadChannel final : public DiscoChannel {
+ public:
+  void Send(const TVMArgs& args) { controler_to_worker_.Send(args); }
+  TVMArgs Recv() { return controler_to_worker_.Recv(); }
+  void Reply(const TVMArgs& args) { worker_to_controler_.Send(args); }
+  TVMArgs RecvReply() { return worker_to_controler_.Recv(); }
+
+  ThreadedMessageQueue controler_to_worker_;
+  ThreadedMessageQueue worker_to_controler_;
+};
+
+class ThreadedCommandQueueObj final : public CommandQueueObj {
+ public:
+  explicit ThreadedCommandQueueObj(int n_workers) {
+    for (int i = 0; i < n_workers; ++i) {
+      std::shared_ptr<DiscoThreadChannel> channel = std::make_shared<DiscoThreadChannel>();
+      WorkerZeroData* data = (i == 0) ? &worker_zero_data_ : nullptr;
+      workers_.emplace_back(std::make_unique<DiscoWorker>(i, n_workers, data, channel));
+      channels_.emplace_back(std::move(channel));
```

Comment:
why are we opening channels for each worker? Isn't the controller supposed to communicate with worker 0 only?

Reply:
This is an implementation detail. It is possible to have an impl that have control messages work with all workers. Especially in threaded case. But it does not have to be the only implementation. So this is fine as a first step for threaded.